metrics/src/macros.rs

```diff
@@ -0,0 +1,809 @@
+/// which includes using macros such as `format!` directly at the callsite. String literals are
+/// preferred for performance where possible.
+///
+/// # Example
+/// ```
+/// # use metrics::register_histogram;
+/// # fn main() {
+/// // A basic histogram:
+/// let histogram = register_histogram!("some_metric_name");
+/// histogram.record(1.0);
+///
+/// // Specifying labels inline, including using constants for either the key or value:
+/// let histogram = register_histogram!("some_metric_name", "service" => "http");
+///
+/// const SERVICE_LABEL: &'static str = "service";
+/// const SERVICE_HTTP: &'static str = "http";
+/// let histogram = register_histogram!("some_metric_name", SERVICE_LABEL => SERVICE_HTTP);
+///
+/// // We can also pass labels by giving a vector or slice of key/value pairs.  In this scenario,
+/// // a unit or description can still be passed in their respective positions:
+/// let dynamic_val = "woo";
+/// let labels = [("dynamic_key", format!("{}!", dynamic_val))];
+/// let histogram = register_histogram!("some_metric_name", &labels);
+///
+/// // As mentioned in the documentation, metric names also can be owned strings, including ones
+/// // generated at the callsite via things like `format!`:
+/// let name = String::from("some_owned_metric_name");
+/// let histogram = register_histogram!(name);
+///
+/// let histogram = register_histogram!(format!("{}_via_format", "name"));
+/// # }
+/// ```
+#[macro_export]
+macro_rules! register_histogram {
+    (target: $target:expr, level: $level:expr, $name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {{
+        let metric_key = $crate::key_var!($name $(, $label_key $(=> $label_value)?)*);
+        let metadata = $crate::metadata_var!($target, $level);
+
+        $crate::recorder().register_histogram(&metric_key, metadata)
+    }};
+    (target: $target:expr, $name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {
+        $crate::register_histogram!(target: $target, level: $crate::Level::INFO, $name $(, $label_key $(=> $label_value)?)*)
+    };
+    (level: $level:expr, $name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {
+        $crate::register_histogram!(target: module_path!(), level: $level, $name $(, $label_key$(=> $label_value)?)*)
+    };
+    ($name:expr $(, $label_key:expr $(=> $label_value:expr)?)* $(,)?) => {
+        $crate::register_histogram!(target: module_path!(), level: $crate::Level::INFO, $name $(, $label_key$(=> $label_value)?)*)
+    };
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! describe {
+    ($method:ident, $name:expr, $unit:expr, $description:expr) => {{
+        if let ::core::option::Option::Some(recorder) = $crate::try_recorder() {
+            recorder.$method($name.into(), ::core::option::Option::Some($unit), $description.into())
+        }
+    }};
+    ($method:ident, $name:expr, $description:expr) => {{
+        if let Some(recorder) = $crate::try_recorder() {
```

Comment:
It does, whoops. I'll try write a test involving `#[no_implicit_prelude]` so that me and others wont make this mistake again.

Reply:
https://github.com/metrics-rs/metrics/pull/386/commits/a901339b4c122d0c2486ba7e44031b80e097c786

I've slipped `#![no_implicit_prelude]` into the doc tests.

I think it might be better to write a dedicate test (or even just an example) where we invoke all combinations of all macros with `#![no_implicit_prelude]` and we don't import things like `use std::convert::From` which weakens the test. I can do this in a follow up PR unless you think its important now.